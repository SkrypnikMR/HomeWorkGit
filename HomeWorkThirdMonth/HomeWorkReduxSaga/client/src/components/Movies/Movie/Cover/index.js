import Cover from './Cover';
import { connect } from 'react-redux';
import { updateMovie } from '../../../../store/movies/actions';
import { getMovieField } from '../../../../store/movies/selectors';

const mapStateToProps = (state, props) => ({
    cover: getMovieField(state, props),
});

const mapDispatchToProps = { updateMovie };

export default connect(mapStateToProps, mapDispatchToProps)(Cover);