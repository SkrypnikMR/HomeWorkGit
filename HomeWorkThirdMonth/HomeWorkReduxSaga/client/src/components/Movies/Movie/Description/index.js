import Description from './Description';
import { connect } from 'react-redux';
import { updateMovie } from '../../../../store/movies/actions';
import { getMovieField } from '../../../../store/movies/selectors';

const mapStateToProps = (state, props) => ({
    description: getMovieField(state, props),
});

const mapDispatchToProps = { updateMovie };

export default connect(mapStateToProps, mapDispatchToProps)(Description);