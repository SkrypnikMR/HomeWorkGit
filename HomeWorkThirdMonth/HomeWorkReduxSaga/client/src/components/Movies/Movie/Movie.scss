.movie{
    width: 400px;
    height: 500px;
    justify-self: center;
    margin:10px;
    display: flex;
    align-items: center;
    overflow-y: hidden auto;
    justify-content: center;
    flex-direction: column;
    border: 1px solid black;
    border-radius: 10px;
    transition-duration: .5s;
    color: white;
    img{
        margin-top: -10px;
        height: 70%;
        width: 100%;
    }
    h5, p {
        width: 100%;
        text-align: center;


    }
    p {
           font-size: 20px;
    }
    h5 {
        font-size: 17px;
        z-index: 1000;
    }
     .description__change{
             margin-top: 5px;
             height: 10%;
             width: 100%;
             display: flex;
             align-items: center;
             justify-content: center;
              input {
            height: 40px;
            width: 90%;
            margin: 15px;
            z-index: 1000;
            position: relative;
            overflow-x: hidden;
            border: none;
            color: #f5da0b;
            &::placeholder{
                color:#f5da0b;
                font-weight: bold;
            }
        }
    }
}