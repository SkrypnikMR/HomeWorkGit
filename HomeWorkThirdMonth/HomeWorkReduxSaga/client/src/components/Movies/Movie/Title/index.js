import Title from './Title.jsx';
import { connect } from 'react-redux';
import { deleteMovie, updateMovie } from '../../../../store/movies/actions';
import { getMovieField } from '../../../../store/movies/selectors';

const mapStateToProps = (state, props) => ({
    title: getMovieField(state, props),
});

const mapDispatchToProps = { deleteMovie, updateMovie };

export default connect(mapStateToProps, mapDispatchToProps)(Title);