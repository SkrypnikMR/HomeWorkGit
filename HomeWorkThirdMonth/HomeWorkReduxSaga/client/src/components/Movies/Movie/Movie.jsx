import React, { useState, useEffect, useMemo } from 'react';
import propTypes from 'prop-types';
import './Movie.scss';
import Button from '../../Button';
import Input from '../../Input';
import { defaultImage } from '../../../constants/uiConstants'
import Loader from '../../Loader';
import { NotificationManager } from 'react-notifications';

const Movie = ({ movie, deleteMovie, updateMovie, isLoading }) => {
    const { cover, title, description } = movie;
    const [state, setState] = useState({
        src: cover,
        errored: false,
        title: { isWaitForChange: false },
        cover: { isWaitForChange: false },
        description: { isWaitForChange: false },
        changeParam: '',
        newData: '',
    });
    useEffect(() => {
        setState({ ...state, src: cover })
    }, [cover]);

    const handleDeleteItem = () => deleteMovie(movie.id);
    const onError = () => setState({ ...state, src: defaultImage, errored: true })
    const handleOnDoubleClick = (e) => {
        const key = e.target.getAttribute('name');
        setState((state) => ({ ...state, [key]: { isWaitForChange: !state[key].isWaitForChange } }));
    }
    const handleOnChange = (e) => setState({ ...state, changeParam: e.target.name, newData: e.target.value });

    const handleOnBlur = (e) => {
        if (e.target.value === '') return NotificationManager.error('Please Enter newData for field', 'INPUT ERROR', 2000);
        const { changeParam, newData } = state;
        updateMovie({ changeParam, newData, id: movie.id });
        setState({ ...state, [e.target.name]: { isWaitForChange: false } });
    }
    return isLoading ? <Loader /> : (
        <div className="movie" >
            <div className="movie__header">
                {
                    state.title.isWaitForChange ? (
                        <Input
                            name='title'
                            onBlurInput={handleOnBlur}
                            onChangeInput={handleOnChange}
                            placeholder='enter new title'
                        />
                    )
                        : (
                            <h5 name="title" onDoubleClick={handleOnDoubleClick}>
                                {title}
                            </h5>
                        )
                }
                <Button onButtonClick={handleDeleteItem} title="X" />
            </div>
            {state.cover.isWaitForChange ? (
                <div className="cover__change">
                    <Input
                        name='cover'
                        onBlurInput={handleOnBlur}
                        onChangeInput={handleOnChange}
                        placeholder='enter new cover'
                    />
                </div>
            )
                : <img src={state.src} onError={onError} onDoubleClick={handleOnDoubleClick} name="cover" />}
            {state.description.isWaitForChange ? (
                <div className="description__change">
                    <Input
                        name='description'
                        onBlurInput={handleOnBlur}
                        onChangeInput={handleOnChange}
                        placeholder='enter new description'
                    />
                </div>
            ) : <p onDoubleClick={handleOnDoubleClick} name="description">{description}</p>}
        </div>
    );
}
Movie.propTypes = {
    movie: propTypes.object,
    deleteMovie: propTypes.func,
    updateMovie: propTypes.func,
    isLoading: propTypes.bool,
};
export default Movie;
