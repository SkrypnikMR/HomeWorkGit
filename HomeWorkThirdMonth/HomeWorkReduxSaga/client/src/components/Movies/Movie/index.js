import { connect } from 'react-redux';
import Movie from './Movie.jsx';
import { deleteMovie, updateMovie } from '../../../store/movies/actions';
import { getMovieById, getIsLoading } from '../../../store/movies/selectors';

const mapStateToProps = (state, props) => ({
    movie: getMovieById(state, props),
    isLoading: getIsLoading(state)
});

const mapDispatchToProps = { deleteMovie, updateMovie };

export default connect(mapStateToProps, mapDispatchToProps)(Movie);
