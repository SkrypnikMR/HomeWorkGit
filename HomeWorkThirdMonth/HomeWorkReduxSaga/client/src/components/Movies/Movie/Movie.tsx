import React from 'react';
import Title from './Title';
import Cover from './Cover';
import Description from './Description';
import './Movie.scss';

interface IMovie { id: string;}

const Movie = ({ id }: IMovie) => {
    return (
        <div className="movie" >
            <Title name='title' id={id} />
            <Cover name="cover" id={id} />
            <Description name="description" id={id} />
        </div>
    );
}


export default React.memo(Movie);
