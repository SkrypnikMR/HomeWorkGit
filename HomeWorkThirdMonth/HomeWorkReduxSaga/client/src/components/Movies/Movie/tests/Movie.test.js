import React from 'react';
import { shallow, mount } from 'enzyme';
import Movie from '../Movie';

describe('Movie', () => {
    let props;
    it('Should match snapshot', () => {
        props = {
            movie: {
                id: 1,
                cover: 'cover',
                title: 'title',
                description: 'desc',
            },
            deleteMovie: jest.fn(),
            updateMovie: jest.fn(),
            isLoading: false
        }
        const component = shallow(<Movie {...props} />);
        expect(component.html()).toMatchSnapshot();
    });
    it('Should have Loader', () => {
        props = {
            movie: {
                id: 1,
                cover: 'cover',
                title: 'title',
                description: 'desc',
            },
            deleteMovie: jest.fn(),
            updateMovie: jest.fn(),
            isLoading: true
        }
        const component = mount(<Movie {...props} />);
        expect(component.find('Loader')).toHaveLength(1);
    });
    it('Should have h5', () => {
        props = {
            movie: {
                id: 1,
                cover: 'cover',
                title: 'title',
                description: 'desc',
            },
            deleteMovie: jest.fn(),
            updateMovie: jest.fn(),
            isLoading: false
        }
        const component = mount(<Movie {...props} />);
        expect(component.find('h5')).toHaveLength(1);
    });
    it('Should have img', () => {
        props = {
            movie: {
                id: 1,
                cover: 'cover',
                title: 'title',
                description: 'desc',
            },
            deleteMovie: jest.fn(),
            updateMovie: jest.fn(),
            isLoading: false
        }
        const component = mount(<Movie {...props} />);
        expect(component.find('img')).toHaveLength(1);
    });
    it('Should Input and no have h5', () => {
        props = {
            movie: {
                id: 1,
                cover: 'cover',
                title: 'title',
                description: 'desc',
            },
            deleteMovie: jest.fn(),
            updateMovie: jest.fn(),
            isLoading: false
        }
        const component = mount(<Movie {...props} />);
        component.find('h5').simulate('dblclick');
        expect(component.find('Input')).toHaveLength(1);
        expect(component.find('h5')).toHaveLength(0);
    });
    it('Should Input and no have img', () => {
        props = {
            movie: {
                id: 1,
                cover: 'cover',
                title: 'title',
                description: 'desc',
            },
            deleteMovie: jest.fn(),
            updateMovie: jest.fn(),
            isLoading: false
        }
        const component = mount(<Movie {...props} />);
        component.find('img').simulate('dblclick');
        expect(component.find('Input')).toHaveLength(1);
        expect(component.find('img')).toHaveLength(0);
    });
    it('Should Input and no have p', () => {
        props = {
            movie: {
                id: 1,
                cover: 'cover',
                title: 'title',
                description: 'desc',
            },
            deleteMovie: jest.fn(),
            updateMovie: jest.fn(),
            isLoading: false
        }
        const component = mount(<Movie {...props} />);
        component.find('p').simulate('dblclick');
        expect(component.find('Input')).toHaveLength(1);
        expect(component.find('p')).toHaveLength(0);
    });
    it('Should Input and no have p', () => {
        props = {
            movie: {
                id: 1,
                cover: 'cover',
                title: 'title',
                description: 'desc',
            },
            deleteMovie: jest.fn(),
            updateMovie: jest.fn(),
            isLoading: false
        }
        const component = mount(<Movie {...props} />);
        component.find('Button').simulate('click');
        expect(props.deleteMovie).toHaveBeenCalledWith(props.movie.id)
    });
    it('Should call onError', () => {
        props = {
            movie: {
                id: 1,
                cover: 'cover',
                title: 'title',
                description: 'desc',
            },
            deleteMovie: jest.fn(),
            updateMovie: jest.fn(),
            isLoading: false
        }
        const component = mount(<Movie {...props} />);
        component.find('img').props().onError();
    });
    it('Should call handleOnChange', () => {
        props = {
            movie: {
                id: 1,
                cover: 'cover',
                title: 'title',
                description: 'desc',
            },
            deleteMovie: jest.fn(),
            updateMovie: jest.fn(),
            isLoading: false
        }
        const component = mount(<Movie {...props} />);
        component.find('img').simulate('dblclick');
        component.find('Input').simulate('change', { taget: { name: 'cover', value: 'new Cover' } })
    });
    it('Should call handleOnBlur but value ="" ', () => {
        props = {
            movie: {
                id: 1,
                cover: 'cover',
                title: 'title',
                description: 'desc',
            },
            deleteMovie: jest.fn(),
            updateMovie: jest.fn(),
            isLoading: false
        }
        const component = mount(<Movie {...props} />);
        component.find('img').simulate('dblclick');
        component.find('Input').simulate('change', { taget: { name: 'cover', value: 'new Cover' } })
        component.find('Input').simulate('blur', { taget: { name: 'cover', value: '' } })
    });
    it('Should call handleOnBlur value = someValue ', () => {
        props = {
            movie: {
                id: 1,
                cover: 'cover',
                title: 'title',
                description: 'desc',
            },
            deleteMovie: jest.fn(),
            updateMovie: jest.fn(),
            isLoading: false
        }
        const component = mount(<Movie {...props} />);
        component.find('img').simulate('dblclick');
        component.find('Input').simulate('change', { taget: { name: 'cover', value: 'new Cover' } })
        component.find('Input').props().onBlurInput({ target: { value: '12321' } })
        expect(props.updateMovie).toHaveBeenCalledWith({ changeParam: 'cover', id: 1, newData: "" });
    });
});
