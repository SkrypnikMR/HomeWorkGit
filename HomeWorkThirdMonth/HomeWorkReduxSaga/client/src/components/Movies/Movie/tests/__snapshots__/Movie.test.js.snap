// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie Should match snapshot 1`] = `"<div class=\\"movie\\"><div class=\\"movie__header\\"><h5 name=\\"title\\">title</h5><button class=\\"custom__button\\">X</button></div><img src=\\"cover\\" name=\\"cover\\"/><p name=\\"description\\">desc</p></div>"`;
